File: android/pulse-core/src/test/kotlin/studio/pulsecircle/pulse/core/testkit/TestDoubles.kt
package studio.pulsecircle.pulse.core.testkit

import studio.pulsecircle.pulse.core.HttpRequest
import studio.pulsecircle.pulse.core.HttpResult
import studio.pulsecircle.pulse.core.PulseCancellable
import studio.pulsecircle.pulse.core.PulseClock
import studio.pulsecircle.pulse.core.PulseExecutor
import studio.pulsecircle.pulse.core.PulseKeyValueStorage
import studio.pulsecircle.pulse.core.PulseLogger
import studio.pulsecircle.pulse.core.PulseQueueStorage
import studio.pulsecircle.pulse.core.PulseTransport

/** Immediate synchronous executor: deterministic single-threaded test runs. */
class ImmediateExecutor : PulseExecutor {
    override fun execute(work: () -> Unit) = work()
}

/**
 * Virtual clock per FIXTURES.md: starts at 2026-01-01T00:00:00.000Z and only
 * moves via [advance], which fires due timers in chronological (then
 * scheduling) order.
 */
class VirtualClock(startMs: Long = FIXTURE_EPOCH_MS) : PulseClock {

    companion object {
        /** 2026-01-01T00:00:00.000Z */
        const val FIXTURE_EPOCH_MS = 1_767_225_600_000L
    }

    private class Timer(val due: Long, val seq: Long, val work: () -> Unit) {
        var cancelled = false
    }

    private var now = startMs
    private var nextSeq = 0L
    private val timers = mutableListOf<Timer>()

    override fun nowMs(): Long = now

    override fun schedule(afterMs: Long, work: () -> Unit): PulseCancellable {
        val timer = Timer(now + afterMs.coerceAtLeast(0), nextSeq++, work)
        timers.add(timer)
        return PulseCancellable { timer.cancelled = true }
    }

    fun advance(ms: Long) {
        val target = now + ms
        while (true) {
            timers.removeAll { it.cancelled }
            val next = timers
                .filter { it.due <= target }
                .minWithOrNull(compareBy({ it.due }, { it.seq }))
                ?: break
            timers.remove(next)
            if (next.due > now) now = next.due
            next.work()
        }
        now = target
    }

    /** Simulates process death: pending timers die with the process. */
    fun cancelAllTimers() {
        timers.clear()
    }
}

/**
 * Mock transport per FIXTURES.md: captures outgoing requests into a FIFO of
 * pending requests; the client's in-flight request stays unresolved until a
 * fixture step (or test) explicitly responds.
 */
class MockTransport : PulseTransport {
    class Pending(val request: HttpRequest, val callback: (HttpResult) -> Unit)

    val pending = ArrayDeque<Pending>()

    override fun send(request: HttpRequest, callback: (HttpResult) -> Unit) {
        pending.addLast(Pending(request, callback))
    }

    fun takeOldest(): Pending? = pending.removeFirstOrNull()
}

class InMemoryKeyValueStorage : PulseKeyValueStorage {
    val map = LinkedHashMap<String, String>()
    override fun get(key: String): String? = map[key]
    override fun set(key: String, value: String) {
        map[key] = value
    }

    override fun remove(key: String) {
        map.remove(key)
    }
}

class InMemoryQueueStorage : PulseQueueStorage {
    val items = mutableListOf<String>()

    override fun loadAll(): List<String> = items.toList()

    override fun append(itemJson: String) {
        items.add(itemJson)
    }

    override fun markConsumed(count: Int) {
        repeat(count.coerceIn(0, items.size)) { items.removeAt(0) }
    }

    override fun replaceAll(items: List<String>) {
        this.items.clear()
        this.items.addAll(items)
    }
}

class RecordingLogger : PulseLogger {
    val debugMessages = mutableListOf<String>()
    val errorMessages = mutableListOf<String>()

    override fun debug(message: String) {
        debugMessages.add(message)
    }

    override fun error(message: String) {
        errorMessages.add(message)
    }
}

File: ios/Tests/PulseSDKTests/TestDoubles.swift
import Foundation
@testable import PulseSDK

// MARK: - Executor

/// Runs work synchronously and immediately: with this executor the client is
/// fully deterministic and every public call settles before returning.
final class ImmediateExecutor: PulseExecutor {
    func execute(_ work: @escaping () -> Void) {
        work()
    }
}

// MARK: - Virtual clock

/// Deterministic clock starting at 2026-01-01T00:00:00.000Z. Time only moves
/// via `advance(ms:)`, which fires due timers in chronological order
/// (including timers scheduled while advancing, if they come due).
final class VirtualClock: PulseClock {

    static let startEpochMs: Int64 = 1_767_225_600_000 // 2026-01-01T00:00:00.000Z

    private(set) var currentMs: Int64 = VirtualClock.startEpochMs

    private struct ScheduledTimer {
        let id: Int
        let dueMs: Int64
        let sequence: Int
        let work: () -> Void
    }

    private var timers: [ScheduledTimer] = []
    private var nextId = 1
    private var sequenceCounter = 0

    func nowMs() -> Int64 {
        return currentMs
    }

    func schedule(afterMs: Int64, _ work: @escaping () -> Void) -> PulseCancellable {
        let id = nextId
        nextId += 1
        sequenceCounter += 1
        timers.append(ScheduledTimer(
            id: id,
            dueMs: currentMs + max(0, afterMs),
            sequence: sequenceCounter,
            work: work
        ))
        return VirtualClockCancellable { [weak self] in
            guard let self = self else { return }
            self.timers.removeAll { $0.id == id }
        }
    }

    func advance(ms: Int64) {
        let target = currentMs + ms
        while true {
            var earliest: ScheduledTimer?
            for timer in timers where timer.dueMs <= target {
                if let best = earliest {
                    if (timer.dueMs, timer.sequence) < (best.dueMs, best.sequence) {
                        earliest = timer
                    }
                } else {
                    earliest = timer
                }
            }
            guard let fire = earliest else { break }
            timers.removeAll { $0.id == fire.id }
            if fire.dueMs > currentMs {
                currentMs = fire.dueMs
            }
            fire.work()
        }
        currentMs = target
    }
}

final class VirtualClockCancellable: PulseCancellable {
    private let onCancel: () -> Void

    init(onCancel: @escaping () -> Void) {
        self.onCancel = onCancel
    }

    func cancel() {
        onCancel()
    }
}

// MARK: - Mock transport

struct MockNetworkError: Error {}

/// FIFO of pending requests. Each `expectRequest` step pops the oldest,
/// matches it, and only then delivers the scripted response; the client's
/// in-flight request stays unresolved until then.
final class MockTransport: PulseTransport {

    struct Pending {
        let request: PulseHTTPRequest
        let completion: (Result<PulseHTTPResponse, Error>) -> Void
    }

    private let lock = NSLock()
    private var pending: [Pending] = []

    func send(_ request: PulseHTTPRequest, completion: @escaping (Result<PulseHTTPResponse, Error>) -> Void) {
        lock.lock()
        pending.append(Pending(request: request, completion: completion))
        lock.unlock()
    }

    func popOldest() -> Pending? {
        lock.lock()
        defer { lock.unlock() }
        if pending.isEmpty {
            return nil
        }
        return pending.removeFirst()
    }

    func peekOldest() -> Pending? {
        lock.lock()
        defer { lock.unlock() }
        return pending.first
    }

    var pendingCount: Int {
        lock.lock()
        defer { lock.unlock() }
        return pending.count
    }
}

// MARK: - In-memory storage

final class InMemoryKeyValueStorage: PulseKeyValueStorage {
    private var values: [String: String] = [:]

    func get(_ key: String) -> String? {
        return values[key]
    }

    func set(_ key: String, _ value: String) {
        values[key] = value
    }

    func remove(_ key: String) {
        values[key] = nil
    }
}

final class InMemoryQueueStorage: PulseQueueStorage {
    private(set) var items: [String] = []

    func loadAll() -> [String] {
        return items
    }

    func append(_ itemJson: String) {
        items.append(itemJson)
    }

    func markConsumed(count: Int) {
        let removeCount = min(max(0, count), items.count)
        if removeCount > 0 {
            items.removeFirst(removeCount)
        }
    }

    func replaceAll(_ newItems: [String]) {
        items = newItems
    }
}

// MARK: - Recording logger

final class RecordingLogger: PulseLogger {
    private let lock = NSLock()
    private(set) var debugMessages: [String] = []
    private(set) var errorMessages: [String] = []

    func log(_ level: PulseLogLevel, _ message: String) {
        lock.lock()
        defer { lock.unlock() }
        switch level {
        case .debug:
            debugMessages.append(message)
        case .error:
            errorMessages.append(message)
        }
    }
}

// MARK: - Seeded random source (SplitMix64)

final class SeededRandomSource: PulseRandomSource {
    private var state: UInt64

    init(seed: UInt64) {
        self.state = seed
    }

    private func next() -> UInt64 {
        state = state &+ 0x9E37_79B9_7F4A_7C15
        var z = state
        z = (z ^ (z >> 30)) &* 0xBF58_476D_1CE4_E5B9
        z = (z ^ (z >> 27)) &* 0x94D0_49BB_1331_11EB
        return z ^ (z >> 31)
    }

    func nextBytes(_ count: Int) -> [UInt8] {
        var out = [UInt8]()
        out.reserveCapacity(count)
        var buffer: UInt64 = 0
        var available = 0
        for _ in 0..<count {
            if available == 0 {
                buffer = next()
                available = 8
            }
            out.append(UInt8(truncatingIfNeeded: buffer))
            buffer >>= 8
            available -= 1
        }
        return out
    }

    func nextUniform() -> Double {
        // 53 high-quality bits -> [0, 1)
        return Double(next() >> 11) * (1.0 / 9_007_199_254_740_992.0)
    }
}
